mantle/util/retry.go

```diff
@@ -72,6 +72,32 @@ func RetryUntilTimeout(timeout, delay time.Duration, f func() error) error {
 	return nil
 }
 
+// Like RetryUntilTimeout, however we will only execute f retries number of times.
+// We will also only execute f async, so that we may interrupt execution when needed.
+func RetryWithTimeoutAsync(timeout time.Duration, retries int, delay time.Duration, f func() error) error {
+	after := time.After(timeout)
+	done := make(chan bool)
+	var err error
+	wrapper := func() {
+		err = f()
+		done <- true
+	}
+	for i := 0; i < retries; i++ {
+		go wrapper()
+		select {
+		case <-after:
+			return fmt.Errorf("time limit exceeded: %v", err)
+		case <-done:
+		}
+
+		if err == nil {
+			break
+		}
+		time.Sleep(delay)
+	}
+	return nil
```

Comment:
If we run out of tries but not out of time, we return `nil` anyway.  We should return the last error instead.